evm/vm/forks/sharding/log_handler.py

```diff
@@ -0,0 +1,99 @@
+import logging
+
+from eth_utils import (
+    to_dict,
+)
+
+
+class LogHandler(object):
+
+    logger = logging.getLogger("evm.chain.sharding.LogHandler")
+
+    def __init__(self, w3, history_size=256):
+        self.history_size = history_size
+        self.w3 = w3
+        # ----------> higher score
+        self.recent_block_hashes = self.get_recent_block_hashes()
+
+    def get_recent_block_hashes(self):
+        block = self.w3.eth.getBlock('latest')
+        recent_hashes = [block['hash']]
+        # initialize the list of recent hashes
+        for _ in range(self.history_size - 1):
+            # break the loop if we hit the genesis block.
+            if block['number'] == 0:
+                break
+            block = self.w3.eth.getBlock(block['parentHash'])
+            recent_hashes.append(block['hash'])
+        reversed_recent_hashes = tuple(reversed(recent_hashes))
+        return reversed_recent_hashes
+
+    def check_chain_head(self):
```

Comment:
Have done it. Thank you for the wise instruction.
May I ask how can I determine which functions should stay in the class, while the others should be put outside the class `LogHandler`? Is the reason that `get_new_logs` stays in the class because it is the only interface used by `VMCHandler`?
Is the reason `check_chain_head`, `get_recent_block_hashes`, `preprocess_block_param` can be put outside the class `LogHandler` because they only depend on `web3` and can be used directly without the logics of logs?
Sorry for asking this newbie question. Just want to know what is the better criteria to design the class.

Reply:
No need to apologize and I'm all for asking questions of any kind.

My general rule of thumb is the following loose collection of axioms and guidelines.

- classes are typically stateful and statefullness is hard to manage.
- functions (especially pure ones) are easier to reason about than classes.
- a class with only one function is just a function disguised as a class.  This also regularly applies to classes with only 2-3 functions.
- a method that doesn't use `self` should almost always be a function.
- it is *ok* to include convenience functions using `@staticmethod`, but normally should only be done if the convenience is *external*, meaning that outside actors who use this class almost always need the convenience function.  If it's only needed internally, it should just be used as a stand-alone function.

Reply:
Thank you so much for giving me these guidelines! They are really helpful.
Just want to conclude,
is it correct that: In general, we should use functions over classes(as pointed in https://github.com/pipermerriam/ethereum-dev-tactical-manual#the-zen-of-the-python-ethereum-team). And it seems the right timing to choose a class over functions is when we have to maintain states besides a set of functions. (and should move functions which is not necessarily resided inside the class to stand-alone functions if possible?)